src/lem/circuit.rs

```diff
@@ -708,13 +734,85 @@ impl Func {
                         unhash_helper!(preimg, img, SlotType::Hash4);
                     }
                     Op::Null(tgt, tag) => {
+                        let tag = g.global_allocator.get_or_alloc_const(cs, tag.to_field())?;
+                        let zero = g.global_allocator.get_or_alloc_const(cs, F::ZERO)?;
+                        let allocated_ptr = AllocatedPtr::from_parts(tag, zero);
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Lit(tgt, lit) => {
+                        let lit_ptr = lit.to_ptr(g.store);
+                        let lit_tag = lit_ptr.tag().to_field();
+                        let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
+                        let allocated_tag = g.global_allocator.get_or_alloc_const(cs, lit_tag)?;
+                        let allocated_hash = g.global_allocator.get_or_alloc_const(cs, lit_hash)?;
+                        let allocated_ptr = AllocatedPtr::from_parts(allocated_tag, allocated_hash);
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Cast(tgt, tag, src) => {
+                        let src = bound_allocations.get(src)?;
+                        let tag = g.global_allocator.get_or_alloc_const(cs, tag.to_field())?;
+                        let allocated_ptr = AllocatedPtr::from_parts(tag, src.hash().clone());
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Add(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        // TODO check that the tags are correct
```

Comment:
It's not as simple as checking for correctness. Depending on the types, this needs to be handled differently — same for all (or at least many) of the other binary operations.

Reply:
Right. But I still don't know whether the checking of tags and the handling of operations should be done by the LEM programmer or by LEM itself. I tend to prefer the low level approach of having less primitives there are simple and it's up to the programmer to implement stuff correctly.